module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -71,6 +80,346 @@ public function __construct(
+     * @return array     Associative array with two keys: present and missing
+     */
+    public function checkForTags($ids)
+    {
+        // Set up return arrays:
+        $retVal = ['present' => [], 'missing' => []];
+
+        // Look up IDs in the table:
+        $dql = 'SELECT IDENTITY(rt.tag) FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'WHERE rt.tag IN (:ids)';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(compact('ids'));
+        $results = $query->getSingleColumnResult();
+
+        // Record all IDs that are present:
+        $retVal['present'] = array_unique($results);
+
+        // Detect missing IDs:
+        $retVal['missing'] = array_diff($ids, $retVal['present']);
+
+        // Send back the results:
+        return $retVal;
+    }
+
+    /**
+     * Get resources associated with a particular tag.
+     *
+     * @param string $tag  Tag to match
+     * @param string $user ID of user owning favorite list
+     * @param string $list ID of list to retrieve (null for all favorites)
+     *
+     * @return array
+     */
+    public function getResourceIDsForTag($tag, $user, $list = null)
+    {
+        $dql = 'SELECT DISTINCT(rt.resource) AS resource_id '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'WHERE ' . ($this->caseSensitive ? 't.tag = :tag' : 'LOWER(t.tag) = LOWER(:tag) ')
+            . 'AND rt.user = :user ';
+
+        $user = is_object($user) ? $user : $this->entityManager->getReference(User::class, $user);
+        $parameters = compact('tag', 'user');
+        if (null !== $list) {
+            $list = is_object($list) ? $list : $this->entityManager->getReference(UserList::class, $list);
+            $dql .= 'AND rt.list = :list';
+            $parameters['list'] = $list;
+        }
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result =  $query->getSingleColumnResult();
+        return $result;
+    }
+
+    /**
+     * Get lists associated with a particular tag.
+     *
+     * @param string|array      $tag        Tag to match
+     * @param null|string|array $listId     List ID to retrieve (null for all)
+     * @param bool              $publicOnly Whether to return only public lists
+     * @param bool              $andTags    Use AND operator when filtering by tag.
+     *
+     * @return array
+     */
+    public function getListsForTag(
+        $tag,
+        $listId = null,
+        $publicOnly = true,
+        $andTags = true
+    ) {
+        $tag = (array)$tag;
+        $listId = $listId ? (array)$listId : null;
+        $dql = 'SELECT IDENTITY(rt.list) as list, COUNT(DISTINCT(rt.tag)) AS tagcnt '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'JOIN rt.list l '
+            . 'WHERE rt.resource IS NULL ';
```

Comment:
The original code includes this, which doesn't seem to have been ported over:

```
            // Restrict to tags by list owner
            $select->where->and->equalTo(
                'resource_tags.user_id',
                new Expression('l.user_id')
            );
```

I think you could just add `AND rt.user_id = l.user_id` or something similar to cover this situation.

I can't actually think of a strong reason why this is necessary -- by the nature of the system, users can only add tags to their own lists at present, and if we added shared lists in the future, we probably would NOT want this restriction... so maybe it's fine to leave it out... but I just wanted to check and see if you had already thought this through, or if it had been dropped by accident.

Reply:
That's the exact reason why I removed it, I just added it back.

Reply:
Yes, it's probably worth removing in future, but best to do that after migration so we don't confuse ourselves if there's some factor we're overlooking.